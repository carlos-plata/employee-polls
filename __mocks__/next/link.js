import React from 'react';

const Link = ({ children, href, ...props }) => {
  return React.cloneElement(React.Children.only(children), { 
    href, 
    ...props
  });
};

Link.displayName = 'Link';
export default Link;
