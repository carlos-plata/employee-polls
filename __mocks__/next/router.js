export const useRouter = jest.fn().mockReturnValue({
    push: jest.fn(),
    replace: jest.fn(),
    prefetch: jest.fn(),
    query: {},
    asPath: '/',
    pathname: '/',
    events: {
      on: jest.fn(),
      off: jest.fn(),
      emit: jest.fn()
    }
  });
    