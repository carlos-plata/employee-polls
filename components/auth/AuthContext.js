import { createContext, useContext, useEffect } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { login } from '../../store/thunks/authThunks';
import { logout } from '../../store/slices/authSlice'; // Import from slice instead of thunk

const AuthContext = createContext();

export function AuthProvider({ children }) {
  const dispatch = useDispatch();
  const { user, isAuthenticated, error, loading } = useSelector((state) => state.auth);
  
  useEffect(() => {
    // Check for existing user in localStorage
    const storedUser = localStorage.getItem('employeePollsUser');
    if (storedUser) {
      try {
        const parsedUser = JSON.parse(storedUser);
        dispatch({ type: 'auth/loginSuccess', payload: parsedUser });
      } catch (error) {
        console.error('Failed to parse stored user:', error);
        localStorage.removeItem('employeePollsUser');
      }
    }
  }, [dispatch]);
  
  // Update localStorage when user changes
  useEffect(() => {
    if (user) {
      localStorage.setItem('employeePollsUser', JSON.stringify(user));
    } else {
      localStorage.removeItem('employeePollsUser');
    }
  }, [user]);
  
  const handleLogin = async (username, password) => {
    try {
      return await dispatch(login(username, password));
    } catch (error) {
      console.error('Login failed:', error);
      throw error;
    }
  };
  
  const handleLogout = () => {
    dispatch(logout());
  };
  
  const value = {
    user,
    isAuthenticated,
    loading,
    error,
    login: handleLogin,
    logout: handleLogout,
  };
  
  return (
    <AuthContext.Provider value={value}>
      {children}
    </AuthContext.Provider>
  );
}

export function useAuth() {
  return useContext(AuthContext);
}
