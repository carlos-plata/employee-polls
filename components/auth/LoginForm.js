import { useState } from 'react';
import { useRouter } from 'next/router';
import { useAuth } from './AuthContext';

export default function LoginForm() {
  const [username, setUsername] = useState('');
  const [password, setPassword] = useState('');
  const [error, setError] = useState('');
  const [isLoading, setIsLoading] = useState(false);
  
  const { login } = useAuth();
  const router = useRouter();
  
  const handleSubmit = async (e) => {
    e.preventDefault();
    
    if (!username || !password) {
      setError('Please enter both username and password');
      return;
    }
    
    setIsLoading(true);
    setError('');
    
    try {
      await login(username, password);
      router.push('/');
    } catch (error) {
      setError(error.message || 'Invalid username or password');
    } finally {
      setIsLoading(false);
    }
  };
  
  return (
    <form className="mt-8 space-y-6" onSubmit={handleSubmit}>
      {error && (
        <div className="bg-red-100 border-l-4 border-red-500 text-red-700 p-4 mb-4" role="alert">
          <p>{error}</p>
        </div>
      )}
      
      <div>
        <label htmlFor="username" className="block text-sm font-medium text-gray-700">
          User
        </label>
        <input
          id="username"
          name="username"
          type="text"
          value={username}
          onChange={(e) => setUsername(e.target.value)}
          required
          className="mt-1 block w-full px-3 py-2 border border-gray-300 rounded-md shadow-sm focus:outline-none focus:ring-blue-500 focus:border-blue-500"
        />
      </div>
      
      <div>
        <label htmlFor="password" className="block text-sm font-medium text-gray-700">
          Password
        </label>
        <input
          id="password"
          name="password"
          type="password"
          value={password}
          onChange={(e) => setPassword(e.target.value)}
          required
          className="mt-1 block w-full px-3 py-2 border border-gray-300 rounded-md shadow-sm focus:outline-none focus:ring-blue-500 focus:border-blue-500"
        />
      </div>
      
      <div>
        <button
          type="submit"
          disabled={isLoading}
          className="w-full flex justify-center py-2 px-4 border border-transparent rounded-md shadow-sm text-sm font-medium text-white bg-blue-600 hover:bg-blue-700 focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-blue-500 disabled:opacity-50"
        >
          {isLoading ? 'Logging in...' : 'Submit'}
        </button>
      </div>
    </form>
  );
}