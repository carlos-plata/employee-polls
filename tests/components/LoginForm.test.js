import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import LoginForm from '../../components/auth/LoginForm';
import { useAuth } from '../../components/auth/AuthContext';
import { useRouter } from 'next/router';

// Mock the useAuth hook
jest.mock('../../components/auth/AuthContext');

// Mock the useRouter hook
jest.mock('next/router', () => ({
  useRouter: jest.fn()
}));

describe('LoginForm', () => {
  // Setup default mocks
  beforeEach(() => {
    useAuth.mockReturnValue({
      login: jest.fn().mockResolvedValue({}),
    });
    
    useRouter.mockReturnValue({
      push: jest.fn(),
    });
  });
  
  // Test 5: Snapshot test for LoginForm
  it('matches snapshot', () => {
    const { container } = render(<LoginForm />);
    expect(container).toMatchSnapshot();
  });
  
  // Test 6: Form validation - check that error message appears when submitting empty form
  it('shows error message when submitting with empty fields', () => {
    // Arrange
    render(<LoginForm />);
    
    // Act
    fireEvent.click(screen.getByRole('button', { name: /submit/i }));
    
    // Assert
    expect(screen.getByText(/please enter both username and password/i)).toBeInTheDocument();
  });
  
  // Test 7: Successful login redirects to home page
  it('redirects to home page on successful login', async () => {
    // Arrange
    const mockLogin = jest.fn().mockResolvedValue({});
    useAuth.mockReturnValue({ login: mockLogin });
    
    const mockPush = jest.fn();
    useRouter.mockReturnValue({ push: mockPush });
    
    render(<LoginForm />);
    
    // Act
    fireEvent.change(screen.getByLabelText(/user/i), { target: { value: 'sarahedo' } });
    fireEvent.change(screen.getByLabelText(/password/i), { target: { value: 'password123' } });
    fireEvent.click(screen.getByRole('button', { name: /submit/i }));
    
    // Assert
    // Wait for the async login to complete
    await screen.findByRole('button', { name: /submit/i });
    expect(mockLogin).toHaveBeenCalledWith('sarahedo', 'password123');
    expect(mockPush).toHaveBeenCalledWith('/');
  });
  
  // Test 8: Shows error message on login failure
  it('shows error message on login failure', async () => {
    // Arrange
    const mockLogin = jest.fn().mockRejectedValue(new Error('Invalid username or password'));
    useAuth.mockReturnValue({ login: mockLogin });
    
    render(<LoginForm />);
    
    // Act
    fireEvent.change(screen.getByLabelText(/user/i), { target: { value: 'wronguser' } });
    fireEvent.change(screen.getByLabelText(/password/i), { target: { value: 'wrongpass' } });
    fireEvent.click(screen.getByRole('button', { name: /submit/i }));
    
    // Assert
    // Wait for the async login to fail
    const errorMsg = await screen.findByText(/Invalid username or password/i);
    expect(errorMsg).toBeInTheDocument();
  });
});
