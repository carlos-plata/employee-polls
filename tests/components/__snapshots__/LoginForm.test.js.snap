// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginForm matches snapshot 1`] = `
<div>
  <form
    class="mt-8 space-y-6"
  >
    <div>
      <label
        class="block text-sm font-medium text-gray-700"
        for="username"
      >
        User
      </label>
      <input
        class="mt-1 block w-full px-3 py-2 border border-gray-300 rounded-md shadow-sm focus:outline-none focus:ring-blue-500 focus:border-blue-500"
        id="username"
        name="username"
        required=""
        type="text"
        value=""
      />
    </div>
    <div>
      <label
        class="block text-sm font-medium text-gray-700"
        for="password"
      >
        Password
      </label>
      <input
        class="mt-1 block w-full px-3 py-2 border border-gray-300 rounded-md shadow-sm focus:outline-none focus:ring-blue-500 focus:border-blue-500"
        id="password"
        name="password"
        required=""
        type="password"
        value=""
      />
    </div>
    <div>
      <button
        class="w-full flex justify-center py-2 px-4 border border-transparent rounded-md shadow-sm text-sm font-medium text-white bg-blue-600 hover:bg-blue-700 focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-blue-500 disabled:opacity-50"
        type="submit"
      >
        Submit
      </button>
    </div>
  </form>
</div>
`;
